### ISSUE TITLE
multiple epoch trainging yields poorer results

### ISSUE THREAD

**Issue by winstonquock** [SkymindIO/deeplearning4j#146]
This is decent from https://github.com/SkymindIO/deeplearning4j/issues/123

Using the epoch iterator with epoch = 5
```java
    if (true) {
        var batch = 1
        val iter = ListDataSetIterator(trainingSet.asList(), 500)
        val epochIter = MultipleEpochsIterator(5, iter)
        nn.fit(epochIter)
//        while (iter.hasNext()) {
//            println("================== train batch ${batch++} ==================")
//            nn.fit(iter.next())
//        }
    } else {
        nn.fit(trainingSet)
    }

// eval with the training set
        val eval = evalNN(nn, trainingSet)
        println("training eval stats: ${eval.stats()}")
        println("F1: ${eval.f1()} precision: ${eval.precision()}. recall: ${eval.recall()}. accuracy: ${eval.accuracy()}")
        for (cls in 0..4) {
            println("  CLASS[${cls}]: precision: ${eval.precision(cls)} recall: ${eval.recall(cls)}")
        }

```

The result of the same data set look like
```
predicted
[
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]

  ...
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
  [1.7813002054312927E-4, 1.876631170750593E-4, 1.7750803652857077E-4, 1.8325930231097453E-4, 1.7452290973929904E-4]
  [1.7894440744510307E-4, 1.7207436707405944E-4, 1.7939264116179627E-4, 1.752479828152295E-4, 1.8154387727762453E-4]
] // 5599x5

training eval stats: 

Actual Class 0 was predicted with Predicted 1 with count 622 times
Actual Class 0 was predicted with Predicted 4 with count 893 times
Actual Class 1 was predicted with Predicted 1 with count 738 times
Actual Class 1 was predicted with Predicted 4 with count 1055 times
Actual Class 2 was predicted with Predicted 1 with count 179 times
Actual Class 2 was predicted with Predicted 4 with count 222 times
Actual Class 3 was predicted with Predicted 1 with count 386 times
Actual Class 3 was predicted with Predicted 4 with count 509 times
Actual Class 4 was predicted with Predicted 1 with count 420 times
Actual Class 4 was predicted with Predicted 4 with count 575 times
==========================F1 Scores========================================
 0.13133916443366014
===========================================================================
F1: 0.13133916443366014 precision: 0.09828354932685768. recall: 0.19789802330111236. accuracy: 0.43314376405237404
  CLASS[0]: precision: 0.0 recall: 0.0
  CLASS[1]: precision: 0.31471215351812365 recall: 0.41160066926938094
  CLASS[2]: precision: 0.0 recall: 0.0
  CLASS[3]: precision: 0.0 recall: 0.0
  CLASS[4]: precision: 0.1767055931161647 recall: 0.5778894472361809
```
As seen, *only 2 classes* were predicted.

with epoch = 1 or just the list iterator or full set trainging, the F1 will be 0.2 with all classes predicted reasonably.  Note: the score is evaluated with the same full training set itself and not a separated test set, so it is not an overfitting problem.

Please try with my data set and setup to see if it is reproducible for you.

**Comment by agibsonccc** [comment]
Would you mind posting this under the batch training issue? I'd like to keep that consolidated in one place. I have re opened it. Thanks!
